### PR TITLE
feat(har): Remotely accessible HAR file

### LINKED ISSUE #8355: [Feature] Access HAR Catpture from Remote Playwright Client
Under Playwright Server mode, `context.recoardHar.path` only saves the path on the server, but there is no way to access the HAR from the client.

Would you be open to a PR that saves the har to recordHar.path on the client?

Alternatively, it would be nice to have something like…

```js
context.har: (args: { path?: string }): Promise<HAR> // or Buffer if we don’t want it structured
```

As of now, I’ve worked around this by running a simple
HTTP file server exposing the directory the server is writing to so that the client
can request.

Example Server/Client:

`server.js`:

```js
const { chromium } = require(“playwright”);

(async () => {
  const server = await chromium.launchServer({
    port: 9090,
  });
  console.log(`Playwright Server Listening; PW_SERVER=${server.wsEndpoint()}`);
})()
```

`client.js`:

```js
// Filename: client.js
const { chromium } = require(“playwright”);

(async () => {
  const browser = await chromium.connect(http://process.env.PW_SERVER);
  const context = await browser.newContext({
    recordHar: { path: “example.har” }
  })
  const page = await context.newPage();
  await page.goto(“https://playwright.dev/”);
  await context.close();
  await browser.close();
})()
```

### PR DESCRIPTION
**Pull request by rwoll**
This change ensure's the HAR file is saved at `recordHar.path` on the
client instead of the server.

NB: The goal was to make this change transparent to the user and NOT
introduce any new APIs. Namely, I want to leave the API open for
potential `context.har.start()` and `context.har.stop()`.

This does BREAK servers that expect the HAR to be at the `recordHar.path`
on the server, but I think that's OK since there haven't been reports
of missing HAR on client making me think not many users are getting
HAR with client and server on different hosts anyways.

Closes #8355

### CODE REVIEW

**Comment by rwoll** [comment]
@pavelfeldman / @dgozman: This is ready for review. Thanks!

File: src/server/supplements/har/harTracer.ts [hunk 1]
```diff
@@ -263,6 +267,8 @@ export class HarTracer {
   }
 
   async flush() {
+    if (this._isFlushed) return;
```
Comment (pavelfeldman) [inline src/server/supplements/har/harTracer.ts:270]: nit: we put return on the next line

File: src/protocol/protocol.yml [hunk 2]
```diff
@@ -684,6 +684,10 @@ BrowserContext:
       returns:
         artifact: Artifact
 
+    harExport:
```
Comment (pavelfeldman) [inline src/protocol/protocol.yml:687]: nit: this is ok, but there was nothing wrong with returning optional artifact from close.

File: src/server/browserContext.ts [hunk 3]
```diff
@@ -74,7 +74,8 @@ export abstract class BrowserContext extends SdkObject {
     this._closePromise = new Promise(fulfill => this._closePromiseFulfill = fulfill);
 
     if (this._options.recordHar)
-      this._harTracer = new HarTracer(this, this._options.recordHar);
+      this._harTracer = new HarTracer(this, {...this._options.recordHar, path: path.join(this._browser.options.artifactsDir, `${createGuid}.har`)});
```
Comment (pavelfeldman) [inline src/server/browserContext.ts:77]: I don't think this does what you think this does, which suggests we might be missing a test with 2 hars.

**Review by pavelfeldman — COMMENTED** [review]
Looks good, has a small bug.

File: src/server/browserContext.ts (hunk 3 above)
Comment (rwoll) [reply src/server/browserContext.ts]: Oops! Missed the actual invocation!

**Comment by rwoll** [comment]
Thanks @pavelfeldman! It's all fixed up and I added a test for the duplicate/wrong filename.

The test failed before the change (8/10 times with `--repeat-each`), but now passes 🎉! (I couldn't get a more deterministic failure since once the context is closed, we don't call saveAs again.)

**Review by pavelfeldman — APPROVED** [review]

**Pull request merged**